README.md
Copyrigth 2013 - Itxaka Serrano Garcia <itxakaserrano@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>




pymoviebot-2.py
--------------

pymoviebot is a bot for reddit. It will keep checking the subreddit /r/fullmovierequest for movie requests. 
On a new request it will check the sqlite databse created from /r/fullmoviesonyoutube and /r/fullmoviesonvimeo 
and if something is found it will post the correct link to the post.


How it works
-------------

database.py will crawl both subreddits and store into a sqlite database the whole thing. post ID, title, url,
reddit url and post author is stored. 
The sqlite database provides us with a quick and small response for any query we do, much better than searching
both subreddits everytime there is a new request.

pymoviebot-2.py will check the request subreddit and get all the posts. Then it will check if they 
are already fulfilled and commented by itself and if not, check and post if there is any coincidence in the database.

It will also flair the post automatically with the proper flair.

TODOs
------

* Use more try/Except to catch more exceptions, in case reddit is down for example.
* Check what happens if the search answers 2 rows of data, like a movie and his HD counterpart. Im guessing nothing happens,
  it tries to post the links again but the bot check that he has commented already so it ignores it.


CHANGELOG.md
CHANGELOG
=


Version 0.3 
-
Added the distinguished to the bot comment so ithe original poster will get a  mail.
Turned the database main work into a function
Several fixes

Version 0.2  
-
Added the flair system
Added the origin of the movie

Version 0.1  
-
Basic version, has the main features already


database.py
# -*- coding: utf-8 -*-

"""
database.py
Copyrigth 2013 Itxaka Serrano <itxakaserrano@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

This code will crawl around /r/fullmoviesonvimeo and /r/fullmoviesonyoutube and extract the info to a sqlite database
This code works in conjuntion with pymoviebot.py
"""

Version = 0.2

# TODO: Try/catch if reddit is down or there is no connection
# DONE! logging so we don't print anything to the screen
# TODO: Catch database UNIQUE violation in case an id is duplicated
# TODO: Fix the count at the end. Even if it doesn't insert anything, the numbers don't match. Dont know why.

import praw
import sqlite3
import logging
import config


# logging parameters below
logging.basicConfig(filename=config.path + 'database.log',level=logging.INFO, format='%(asctime)s - %(message)s', datefmt='%d/%m/%Y %H:%M:%S')
logging.info("Program Start")

user_agent = ("pymoviebot 0.1 by /u/itxaka")  # API guidelines enforce this
count = 0

r = praw.Reddit(user_agent=user_agent)

db = sqlite3.connect("data.sql")
c = db.cursor()
c.execute("Create table if not exists movie(id text UNIQUE, name text, url text, author text, reddit_url text, subreddit text, status text)")


def updateDb(reddit):
    global count
    subreddit = r.get_subreddit(reddit)
    for submission in subreddit.get_new(limit=1000):
        c.execute('SELECT * FROM movie where id=?', (submission.id,))
        if c.fetchone():  # check if the id is already in the database
            count += 1  # if it was, add one so we can check the db at the end
        else:
            # use str() for the author as it won't work any other way
            c.execute('INSERT INTO movie VALUES(?,?,?,?,?,?,?)', (submission.id,submission.title, submission.url, str(submission.author), submission.permalink, "fullmoviesonvimeo", "ok"))
            db.commit()

updateDb("fullmoviesonvimeo")
updateDb("fullmoviesonyoutube")

# we will execute a little check in order to see if we found new items
c.execute("SELECT COUNT(*) FROM movie")
controlCheck = c.fetchone()[0]

if controlCheck == count:
    logging.info("Database is up to date")
else:
    logging.info("Database was updated")
    logging.info("There was " + str(count) + " items before. There is " + str(controlCheck) + " items now.")

# Don't forget to close the database (¬_¬)
c.close()
db.close()
logging.info("Program End")


pymoviebot-2.py
# -*- coding: utf-8 -*-

"""
pymoviebot-2.py
Copyrigth 2013 Itxaka Serrano <itxakaserrano@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


This code works in conjuntion with database.py


New features:
- Does not need to be running all the time, has to be scheduled by cron
- Does not rely on a list to check the submissions, now it checks the comments.
- It fills the new flair "Not filled"
- Distinguishes posts

"""


import praw
import time
import sqlite3
import logging
import requests
import config

# logging parameters below
logging.basicConfig(filename=config.path + 'pymoviebot.log', level=logging.INFO, format='%(asctime)s - %(message)s', datefmt='%d/%m/%Y %H:%M:%S')
logging.info("Program Start")

user_agent = "pymoviebot 2.0 by /u/itxaka" # API guidelines suggest this
r = praw.Reddit(user_agent=user_agent)

try:
    r.login(config.user, config.password)  # login so we can post comments!
except (praw.errors.InvalidUser, praw.errors.InvalidUserPass):
    logging.error("User or password is incorrect, exiting...")
    exit(0)
except requests.HTTPError as e:
    logging.error("We got an http error: " + str(e))

db = sqlite3.connect(config.path + "data.sql")
c = db.cursor()

subreddit = r.get_subreddit('fullmovierequest')

try:
    for submission in subreddit.get_new(limit=500):  # we get all we can
        try:
            flat_comments = praw.helpers.flatten_tree(submission.comments)  # flat everything
            for comment in flat_comments:
                if "pymoviebot" in str(comment.author):
                    if "moderator" in str(comment.distinguished):
                        pass
                    else:
                        logging.info("Comment distinguished")
                        comment.distinguish()
        except:
            pass
        else:
            c.execute("SELECT * FROM movie WHERE name LIKE ?", ("%" + submission.title + "%",))
            for row in c.fetchall():  # I still don't know what happens when I get 2 results back...
                a = True  # set this so we can keep retrying the comment and not stop until we post it
                commentAuthors = []  # set this so we check live if we have a comment on the submission
                flat_comments = praw.helpers.flatten_tree(submission.comments)  # flat everything
                for comment in flat_comments:
                    commentAuthors.append(comment.author)  # generate a list of usernames that posted
                if "pymoviebot" in str(commentAuthors):  # check if we are in that list
                        pass
                else:
                    while a:
                        try:
                            z = submission.add_comment("The movie ***" + submission.title + "*** can be found in: [here]("+ row[4] + ") thanks to /u/" + row[3] + "\n\n   *I am a bot. for comments or suggestions write [me](http://www.reddit.com/message/compose/?to=Itxaka)*")
                            # Set up the flair
                            submission_to_flair = r.get_submission(submission_id=submission.id)
                            if row[5] == "fullmoviesonvimeo":
                                subreddit.set_flair(submission_to_flair,"View on Vimeo", "vimeo")
                                logging.info("Flair Vimeo set for: " + str(unicode(submission.title)))
                            elif row[5] == "fullmoviesonyoutube":
                                subreddit.set_flair(submission_to_flair,"View on YouTube", "youtube")
                                logging.info("Flair Youtube set for: " + str(submission.title))
                            else:
                                subreddit.set_flair(submission_to_flair,"Request not fullfilled", "notfilled")
                                logging.info("Flair not filled set for: " + str(submission.title))
                            logging.info("Comment posted for " + str(submission.title))
                            a = False  # as soon as we post a comment we get out of the while
                        except praw.errors.RateLimitExceeded:
                            logging.warning("Rate limit exceeded, trying in a few seconds...")
                            time.sleep(60)  # I believe that the rate is around 8 minutes and decreases with the karma.
except requests.HTTPError as e:
    logging.error("We got an http error: " + str(e))
c.close()  # close cursor
db.close()  # close db
logging.info("Program End")

pymoviebot.py
# -*- coding: utf-8 -*-

"""
pymoviebot.py
Copyrigth 2013 Itxaka Serrano <itxakaserrano@gmail.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


This code works in conjuntion with database.py
"""

# DONE! Add logging so it can provide feedback
# DONE! Put it in a loop so it runs indefinitly? Not sure if we want that and a time.sleep() or we rather launch it every X minutes like with cron
# DONE! Try/except in case reddit is down
# TODO: Check what happens if I get back 2 results when searching for a movie. Maybe I get a normal and HD version and all hell breaks loose?


import praw
import time
import sqlite3
import logging
import requests
import config

# logging parameters below
logging.basicConfig(filename=config.path + 'pymoviebot.log',level=logging.INFO, format='%(asctime)s - %(message)s', datefmt='%d/%m/%Y %H:%M:%S')
logging.info("Program Start")

user_agent = ("pymoviebot 0.3 by /u/itxaka") # API guidelines suggest this
already_done = []  # the ones we have submitted the comment to go here

while True:
    r = praw.Reddit(user_agent=user_agent)
    try:
        r.login(config.user,config.password)  # login so we can post comments!
    except (praw.errors.InvalidUser, praw.errors.InvalidUserPass):
        logging.error("User or password is incorrect, exiting...")
        exit(0)
    except requests.HTTPError as e:
        logging.error("We got an http error: " + str(e))
    db = sqlite3.connect("data.sql")
    c = db.cursor()
    subreddit = r.get_subreddit('fullmovierequest')
    try:
        for submission in subreddit.get_new(limit=500):  # we get all we can
            try:
                flat_comments = praw.helpers.flatten_tree(submission.comments)  # flat everything
                for comment in flat_comments:
                    if "pymoviebot" in str(comment.author):
                        if "moderator" in str(comment.distinguished):
                            pass
                        else:
                            logging.info("Comment distinguished")
                            comment.distinguish()
            except:
                pass
            if submission.id in already_done:  # first check to see if we added it to the done list
                pass
            else:
                c.execute("SELECT * FROM movie WHERE name LIKE ?", ("%" + submission.title + "%",))
                for row in c.fetchall():  # I still don't know what happens when I get 2 results back...
                    already_done.append(submission.id)  # we found it so we add it to the list of dones.
                    a = True  # set this so we can keep retying the comment and not stop until we post it
                    commentAuthors = []  # set this so we check live if we have a comment on the submission
                    flat_comments = praw.helpers.flatten_tree(submission.comments)  # flat everything
                    for comment in flat_comments:
                        commentAuthors.append(comment.author)  # generate a list of usernames that posted
                    if "pymoviebot" in str(commentAuthors):  # check if we are in that list
                            pass
                    else:
                        while a:
                            try:
                                z = submission.add_comment("The movie ***" + submission.title + "*** can be found in: [here]("+ row[4] + ") thanks to /u/" + row[3] + "\n\n   *I am a bot. for comments or suggestions write [me](http://www.reddit.com/message/compose/?to=Itxaka)*")
                                # Set up the flair
                                submission_to_flair = r.get_submission(submission_id=submission.id)
                                if row[5] == "fullmoviesonvimeo":
                                    subreddit.set_flair(submission_to_flair,"View on Vimeo", "vimeo")
                                    logging.info("Flair Vimeo set for: " + str(unicode(submission.title)))
                                else:
                                    subreddit.set_flair(submission_to_flair,"View on YouTube", "youtube")
                                    logging.info("Flair Youtube set for: " + str(submission.title))
                                logging.info("Comment posted for " + str(submission.title))
                                a = False  # as soon as we post a comment we get out of the while
                            except praw.errors.RateLimitExceeded:
                                logging.warning("Rate limit exceeded, trying in a few seconds...")
                                time.sleep(60)  # I believe that the rate is around 8 minutes and decreases with the karma.
    except requests.HTTPError as e:
        logging.error("We got an http error: " + str(e))
    c.close()  # close cursor
    db.close()  # close db
    logging.info("One round completed, starting agian in 5 minutes.")
    time.sleep(300)  # We sleep for 5 minutes before checking again, I believe new submissions arent provided until 30 seconds after being posted
    logging.info("There we go again.")

logging.info("Program End")